File: android/src/main/java/com/guichaguri/trackplayer/service/player/ExoPlayback.java
package com.guichaguri.trackplayer.service.player;

import android.content.Context;
import android.support.v4.media.session.PlaybackStateCompat;
import android.util.Log;
import androidx.annotation.NonNull;
import com.facebook.react.bridge.Promise;
import com.google.android.exoplayer2.*;
import com.google.android.exoplayer2.Player.EventListener;
import com.google.android.exoplayer2.Timeline.Window;
import com.google.android.exoplayer2.metadata.Metadata;
import com.google.android.exoplayer2.metadata.MetadataOutput;
import com.google.android.exoplayer2.metadata.icy.IcyHeaders;
import com.google.android.exoplayer2.metadata.icy.IcyInfo;
import com.google.android.exoplayer2.metadata.id3.TextInformationFrame;
import com.google.android.exoplayer2.metadata.id3.UrlLinkFrame;
import com.google.android.exoplayer2.source.TrackGroup;
import com.google.android.exoplayer2.source.TrackGroupArray;
import com.google.android.exoplayer2.trackselection.TrackSelectionArray;
import com.guichaguri.trackplayer.service.MusicManager;
import com.guichaguri.trackplayer.service.Utils;
import com.guichaguri.trackplayer.service.models.Track;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

/**
 * @author Guichaguri
 */
public abstract class ExoPlayback<T extends Player> implements EventListener, MetadataOutput {

    protected final Context context;
    protected final MusicManager manager;
    protected final T player;

    protected List<Track> queue = Collections.synchronizedList(new ArrayList<>());

    // https://github.com/google/ExoPlayer/issues/2728
    protected int lastKnownWindow = C.INDEX_UNSET;
    protected long lastKnownPosition = C.POSITION_UNSET;
    protected int previousState = PlaybackStateCompat.STATE_NONE;
    protected float volumeMultiplier = 1.0F;

    public ExoPlayback(Context context, MusicManager manager, T player) {
        this.context = context;
        this.manager = manager;
        this.player = player;

        Player.MetadataComponent component = player.getMetadataComponent();
        if(component != null) component.addMetadataOutput(this);
    }

    public void initialize() {
        player.addListener(this);
    }

    public List<Track> getQueue() {
        return queue;
    }

    public abstract void add(Track track, int index, Promise promise);

    public abstract void add(Collection<Track> tracks, int index, Promise promise);

    public abstract void remove(List<Integer> indexes, Promise promise);

    public abstract void removeUpcomingTracks();

    public void updateTrack(int index, Track track) {
        int currentIndex = player.getCurrentWindowIndex();

        queue.set(index, track);

        if(currentIndex == index)
            manager.getMetadata().updateMetadata(this, track);
    }

    public Track getCurrentTrack() {
        int index = player.getCurrentWindowIndex();
        return index < 0 || index >= queue.size() ? null : queue.get(index);
    }

    public void skip(String id, Promise promise) {
        if(id == null || id.isEmpty()) {
            promise.reject("invalid_id", "The ID can't be null or empty");
            return;
        }

        for(int i = 0; i < queue.size(); i++) {
            if(id.equals(queue.get(i).id)) {
                lastKnownWindow = player.getCurrentWindowIndex();
                lastKnownPosition = player.getCurrentPosition();

                player.seekToDefaultPosition(i);
                promise.resolve(null);
                return;
            }
        }

        promise.reject("track_not_in_queue", "Given track ID was not found in queue");
    }

    public void skipToPrevious(Promise promise) {
        int prev = player.getPreviousWindowIndex();

        if(prev == C.INDEX_UNSET) {
            promise.reject("no_previous_track", "There is no previous track");
            return;
        }

        lastKnownWindow = player.getCurrentWindowIndex();
        lastKnownPosition = player.getCurrentPosition();

        player.seekToDefaultPosition(prev);
        promise.resolve(null);
    }

    public void skipToNext(Promise promise) {
        int next = player.getNextWindowIndex();

        if(next == C.INDEX_UNSET) {
            promise.reject("queue_exhausted", "There is no tracks left to play");
            return;
        }

        lastKnownWindow = player.getCurrentWindowIndex();
        lastKnownPosition = player.getCurrentPosition();

        player.seekToDefaultPosition(next);
        promise.resolve(null);
    }

    public void play() {
        player.setPlayWhenReady(true);
    }

    public void pause() {
        player.setPlayWhenReady(false);
    }

    public void stop() {
        lastKnownWindow = player.getCurrentWindowIndex();
        lastKnownPosition = player.getCurrentPosition();

        player.stop(false);
        player.setPlayWhenReady(false);
        player.seekTo(lastKnownWindow,0);
    }

    public void reset() {
        lastKnownWindow = player.getCurrentWindowIndex();
        lastKnownPosition = player.getCurrentPosition();

        player.stop(true);
        player.setPlayWhenReady(false);
    }

    public boolean isRemote() {
        return false;
    }

    public long getPosition() {
        return player.getCurrentPosition();
    }

    public long getBufferedPosition() {
        return player.getBufferedPosition();
    }

    public long getDuration() {
        Track current = getCurrentTrack();

        if (current != null && current.duration > 0) {
            return current.duration;
        }

        long duration = player.getDuration();

        return duration == C.TIME_UNSET ? 0 : duration;
    }

    public void seekTo(long time) {
        lastKnownWindow = player.getCurrentWindowIndex();
        lastKnownPosition = player.getCurrentPosition();

        player.seekTo(time);
    }

    public float getVolume() {
        return getPlayerVolume() / volumeMultiplier;
    }

    public void setVolume(float volume) {
        setPlayerVolume(volume * volumeMultiplier);
    }

    public void setVolumeMultiplier(float multiplier) {
        setPlayerVolume(getVolume() * multiplier);
        this.volumeMultiplier = multiplier;
    }

    public abstract float getPlayerVolume();

    public abstract void setPlayerVolume(float volume);

    public float getRate() {
        return player.getPlaybackParameters().speed;
    }

    public void setRate(float rate) {
        player.setPlaybackParameters(new PlaybackParameters(rate, player.getPlaybackParameters().pitch));
    }

    public int getState() {
        switch(player.getPlaybackState()) {
            case Player.STATE_BUFFERING:
                return player.getPlayWhenReady() ? PlaybackStateCompat.STATE_BUFFERING : PlaybackStateCompat.STATE_CONNECTING;
            case Player.STATE_ENDED:
                return PlaybackStateCompat.STATE_STOPPED;
            case Player.STATE_IDLE:
                return PlaybackStateCompat.STATE_NONE;
            case Player.STATE_READY:
                return player.getPlayWhenReady() ? PlaybackStateCompat.STATE_PLAYING : PlaybackStateCompat.STATE_PAUSED;
        }
        return PlaybackStateCompat.STATE_NONE;
    }

    public void destroy() {
        player.release();
    }

    @Override
    public void onTimelineChanged(@NonNull Timeline timeline, int reason) {
        Log.d(Utils.LOG, "onTimelineChanged: " + reason);

        if((reason == Player.TIMELINE_CHANGE_REASON_PREPARED || reason == Player.TIMELINE_CHANGE_REASON_DYNAMIC) && !timeline.isEmpty()) {
            onPositionDiscontinuity(Player.DISCONTINUITY_REASON_INTERNAL);
        }
    }

    @Override
    public void onPositionDiscontinuity(int reason) {
        Log.d(Utils.LOG, "onPositionDiscontinuity: " + reason);

        if(lastKnownWindow != player.getCurrentWindowIndex()) {
            Track previous = lastKnownWindow == C.INDEX_UNSET || lastKnownWindow >= queue.size() ? null : queue.get(lastKnownWindow);
            Track next = getCurrentTrack();

            // Track changed because it ended
            // We'll use its duration instead of the last known position
            if (reason == Player.DISCONTINUITY_REASON_PERIOD_TRANSITION && lastKnownWindow != C.INDEX_UNSET) {
                if (lastKnownWindow >= player.getCurrentTimeline().getWindowCount()) return;
                long duration = player.getCurrentTimeline().getWindow(lastKnownWindow, new Window()).getDurationMs();
                if(duration != C.TIME_UNSET) lastKnownPosition = duration;
            }

            manager.onTrackUpdate(previous, lastKnownPosition, next);
        }

        lastKnownWindow = player.getCurrentWindowIndex();
        lastKnownPosition = player.getCurrentPosition();
    }

    @Override
    public void onTracksChanged(TrackGroupArray trackGroups, @NonNull TrackSelectionArray trackSelections) {
        for(int i = 0; i < trackGroups.length; i++) {
            // Loop through all track groups.
            // As for the current implementation, there should be only one
            TrackGroup group = trackGroups.get(i);

            for(int f = 0; f < group.length; f++) {
                // Loop through all formats inside the track group
                Format format = group.getFormat(f);

                // Parse the metadata if it is present
                if (format.metadata != null) {
                    onMetadata(format.metadata);
                }
            }
        }
    }

    @Override
    public void onLoadingChanged(boolean isLoading) {
        // Buffering updates
    }

    @Override
    public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
        int state = getState();

        if(state != previousState) {
            if(Utils.isPlaying(state) && !Utils.isPlaying(previousState)) {
                manager.onPlay();
            } else if(Utils.isPaused(state) && !Utils.isPaused(previousState)) {
                manager.onPause();
            } else if(Utils.isStopped(state) && !Utils.isStopped(previousState)) {
                manager.onStop();
            }

            manager.onStateChange(state);
            previousState = state;

            if(state == PlaybackStateCompat.STATE_STOPPED) {
                Track previous = getCurrentTrack();
                long position = getPosition();
                manager.onTrackUpdate(previous, position, null);
                manager.onEnd(previous, position);
            }
        }
    }

    @Override
    public void onRepeatModeChanged(int repeatMode) {
        // Repeat mode update
    }

    @Override
    public void onShuffleModeEnabledChanged(boolean shuffleModeEnabled) {
        // Shuffle mode update
    }

    @Override
    public void onPlayerError(ExoPlaybackException error) {
        String code;

        if(error.type == ExoPlaybackException.TYPE_SOURCE) {
            code = "playback-source";
        } else if(error.type == ExoPlaybackException.TYPE_RENDERER) {
            code = "playback-renderer";
        } else {
            code = "playback"; // Other unexpected errors related to the playback
        }

        manager.onError(code, error.getCause().getMessage());
    }

    @Override
    public void onPlaybackParametersChanged(@NonNull PlaybackParameters playbackParameters) {
        // Speed or pitch changes
    }

    @Override
    public void onSeekProcessed() {
        // Finished seeking
    }

    private void handleId3Metadata(Metadata metadata) {
        String title = null, url = null, artist = null, album = null, date = null, genre = null;

        for(int i = 0; i < metadata.length(); i++) {
            Metadata.Entry entry = metadata.get(i);

            if (entry instanceof TextInformationFrame) {
                // ID3 text tag
                TextInformationFrame id3 = (TextInformationFrame) entry;
                String id = id3.id.toUpperCase();

                if (id.equals("TIT2") || id.equals("TT2")) {
                    title = id3.value;
                } else if (id.equals("TALB") || id.equals("TOAL") || id.equals("TAL")) {
                    album = id3.value;
                } else if (id.equals("TOPE") || id.equals("TPE1") || id.equals("TP1")) {
                    artist = id3.value;
                } else if (id.equals("TDRC") || id.equals("TOR")) {
                    date = id3.value;
                } else if (id.equals("TCON") || id.equals("TCO")) {
                    genre = id3.value;
                }

            } else if (entry instanceof UrlLinkFrame) {
                // ID3 URL tag
                UrlLinkFrame id3 = (UrlLinkFrame) entry;
                String id = id3.id.toUpperCase();

                if (id.equals("WOAS") || id.equals("WOAF") || id.equals("WOAR") || id.equals("WAR")) {
                    url = id3.url;
                }

            }
        }

        if (title != null || url != null || artist != null || album != null || date != null || genre != null) {
            manager.onMetadataReceived("id3", title, url, artist, album, date, genre);
        }
    }

    private void handleIcyMetadata(Metadata metadata) {
        for (int i = 0; i < metadata.length(); i++) {
            Metadata.Entry entry = metadata.get(i);

            if(entry instanceof IcyHeaders) {
                // ICY headers
                IcyHeaders icy = (IcyHeaders)entry;

                manager.onMetadataReceived("icy-headers", icy.name, icy.url, null, null, null, icy.genre);

            } else if(entry instanceof IcyInfo) {
                // ICY data
                IcyInfo icy = (IcyInfo)entry;

                String artist, title;
                int index = icy.title == null ? -1 : icy.title.indexOf(" - ");

                if (index != -1) {
                    artist = icy.title.substring(0, index);
                    title = icy.title.substring(index + 3);
                } else {
                    artist = null;
                    title = icy.title;
                }

                manager.onMetadataReceived("icy", title, icy.url, artist, null, null, null);

            }
        }
    }

    @Override
    public void onMetadata(@NonNull Metadata metadata) {
        handleId3Metadata(metadata);
        handleIcyMetadata(metadata);
    }
}

File: android/src/main/java/com/guichaguri/trackplayer/service/player/ExoPlayback.kt
package com.guichaguri.trackplayer.service.player

import android.content.Context
import android.support.v4.media.session.PlaybackStateCompat
import android.util.Log
import com.facebook.react.bridge.Promise
import com.google.android.exoplayer2.*
import com.google.android.exoplayer2.metadata.Metadata
import com.google.android.exoplayer2.metadata.MetadataOutput
import com.google.android.exoplayer2.source.TrackGroupArray
import com.google.android.exoplayer2.trackselection.TrackSelectionArray
import com.guichaguri.trackplayer.service.MusicManager
import com.guichaguri.trackplayer.service.Utils
import com.guichaguri.trackplayer.service.models.Track
import java.util.*

/**
 * @author Guichaguri
 */
abstract class ExoPlayback<T : Player?>(
    protected val context: Context,
    protected val manager: MusicManager,
    protected val player: T,
    protected var autoUpdateMetadata: Boolean
) : Player.EventListener, MetadataOutput {
    var queue = Collections.synchronizedList(ArrayList<Track>())

    // https://github.com/google/ExoPlayer/issues/2728
    protected var lastKnownWindow = C.INDEX_UNSET
    protected var lastKnownPosition = C.POSITION_UNSET.toLong()
    protected var previousState = PlaybackStateCompat.STATE_NONE
    var volumeMultiplier = 1.0f
        set(value) {
            playerVolume = volume * value
            field = value
        }
    open fun initialize() {
        player!!.addListener(this)
    }

    abstract fun add(track: Track, index: Int, promise: Promise)
    abstract fun add(tracks: Collection<Track>, index: Int, promise: Promise)
    abstract fun remove(indexes: List<Int>, promise: Promise)
    abstract fun removeUpcomingTracks()
    abstract var repeatMode: Int
    fun updateTrack(index: Int, track: Track) {
        val currentIndex = player!!.currentWindowIndex
        queue[index] = track
        if (currentIndex == index) manager.metadata.updateMetadata(this, track)
    }

    val currentTrackIndex: Int?
        get() {
            val index = player!!.currentWindowIndex
            return if (index < 0 || index >= queue.size) null else index
        }
    val currentTrack: Track?
        get() {
            val index = player!!.currentWindowIndex
            return if (index < 0 || index >= queue.size) null else queue[index]
        }

    fun skip(index: Int, promise: Promise) {
        if (index < 0 || index >= queue.size) {
            promise.reject("index_out_of_bounds", "The index is out of bounds")
            return
        }
        lastKnownWindow = player!!.currentWindowIndex
        lastKnownPosition = player.currentPosition
        player.seekToDefaultPosition(index)
        promise.resolve(null)
    }

    fun skipToPrevious(promise: Promise) {
        val prev = player!!.previousWindowIndex
        if (prev == C.INDEX_UNSET) {
            promise.reject("no_previous_track", "There is no previous track")
            return
        }
        lastKnownWindow = player.currentWindowIndex
        lastKnownPosition = player.currentPosition
        player.seekToDefaultPosition(prev)
        promise.resolve(null)
    }

    fun skipToNext(promise: Promise) {
        val next = player!!.nextWindowIndex
        if (next == C.INDEX_UNSET) {
            promise.reject("queue_exhausted", "There is no tracks left to play")
            return
        }
        lastKnownWindow = player.currentWindowIndex
        lastKnownPosition = player.currentPosition
        player.seekToDefaultPosition(next)
        promise.resolve(null)
    }

    open fun play() {
        player!!.playWhenReady = true
    }

    fun pause() {
        player!!.playWhenReady = false
    }

    open fun stop() {
        lastKnownWindow = player!!.currentWindowIndex
        lastKnownPosition = player.currentPosition
        player.stop(false)
        player.playWhenReady = false
        player.seekTo(lastKnownWindow, 0)
    }

    open fun reset() {
        lastKnownWindow = player!!.currentWindowIndex
        lastKnownPosition = player.currentPosition
        player.stop(true)
        player.playWhenReady = false
    }

    val isRemote: Boolean
        get() = false

    fun shouldAutoUpdateMetadata(): Boolean {
        return autoUpdateMetadata
    }

    val position: Long
        get() = player!!.currentPosition
    val bufferedPosition: Long
        get() = player!!.bufferedPosition
    val duration: Long
        get() {
            val current = currentTrack
            if (current != null && current.duration > 0) {
                return current.duration
            }
            val duration = player!!.duration
            return if (duration == C.TIME_UNSET) 0 else duration
        }

    open fun seekTo(time: Long) {
        lastKnownWindow = player!!.currentWindowIndex
        lastKnownPosition = player.currentPosition
        player.seekTo(time)
    }

    var volume: Float
        get() = playerVolume / volumeMultiplier
        set(volume) {
            playerVolume = volume * volumeMultiplier
        }

//    fun setVolumeMultiplier(multiplier: Float) {
//        playerVolume = volume * multiplier
//        volumeMultiplier = multiplier
//    }

    abstract var playerVolume: Float
    var rate: Float
        get() = player!!.playbackParameters.speed
        set(rate) {
            player!!.setPlaybackParameters(
                PlaybackParameters(
                    rate,
                    player.playbackParameters.pitch
                )
            )
        }
    val state: Int
        get() {
            when (player!!.playbackState) {
                Player.STATE_BUFFERING -> return if (player.playWhenReady) PlaybackStateCompat.STATE_BUFFERING else PlaybackStateCompat.STATE_CONNECTING
                Player.STATE_ENDED -> return PlaybackStateCompat.STATE_STOPPED
                Player.STATE_IDLE -> return PlaybackStateCompat.STATE_NONE
                Player.STATE_READY -> return if (player.playWhenReady) PlaybackStateCompat.STATE_PLAYING else PlaybackStateCompat.STATE_PAUSED
            }
            return PlaybackStateCompat.STATE_NONE
        }

    open fun destroy() {
        player!!.release()
    }

    override fun onTimelineChanged(timeline: Timeline, reason: Int) {
        Log.d(Utils.LOG, "onTimelineChanged: $reason")
        if ((reason == Player.TIMELINE_CHANGE_REASON_PREPARED || reason == Player.TIMELINE_CHANGE_REASON_DYNAMIC) && !timeline.isEmpty) {
            onPositionDiscontinuity(Player.DISCONTINUITY_REASON_INTERNAL)
        }
    }

    override fun onPositionDiscontinuity(reason: Int) {
        Log.d(Utils.LOG, "onPositionDiscontinuity: $reason")
        if (lastKnownWindow != player!!.currentWindowIndex) {
            val prevIndex = if (lastKnownWindow == C.INDEX_UNSET) null else lastKnownWindow
            val nextIndex = currentTrackIndex
            val next = if (nextIndex == null) null else queue[nextIndex]

            // Track changed because it ended
            // We'll use its duration instead of the last known position
            if (reason == Player.DISCONTINUITY_REASON_PERIOD_TRANSITION && lastKnownWindow != C.INDEX_UNSET) {
                if (lastKnownWindow >= player.currentTimeline.windowCount) return
                val duration =
                    player.currentTimeline.getWindow(lastKnownWindow, Timeline.Window()).durationMs
                if (duration != C.TIME_UNSET) lastKnownPosition = duration
            }
            manager.onTrackUpdate(prevIndex, lastKnownPosition, nextIndex, next)
        } else if (reason == Player.DISCONTINUITY_REASON_PERIOD_TRANSITION && lastKnownWindow == player.currentWindowIndex) {
            val nextIndex = currentTrackIndex
            val next = if (nextIndex == null) null else queue[nextIndex]
            val duration =
                player.currentTimeline.getWindow(lastKnownWindow, Timeline.Window()).durationMs
            if (duration != C.TIME_UNSET) lastKnownPosition = duration
            manager.onTrackUpdate(nextIndex, lastKnownPosition, nextIndex, next)
        }
        lastKnownWindow = player.currentWindowIndex
        lastKnownPosition = player.currentPosition
    }

    override fun onTracksChanged(
        trackGroups: TrackGroupArray,
        trackSelections: TrackSelectionArray
    ) {
        for (i in 0 until trackGroups.length) {
            // Loop through all track groups.
            // As for the current implementation, there should be only one
            val group = trackGroups[i]
            for (f in 0 until group.length) {
                // Loop through all formats inside the track group
                val format = group.getFormat(f)

                // Parse the metadata if it is present
                if (format.metadata != null) {
                    onMetadata(format.metadata!!)
                }
            }
        }
    }

    override fun onLoadingChanged(isLoading: Boolean) {
        // Buffering updates
    }

    override fun onPlayerStateChanged(playWhenReady: Boolean, playbackState: Int) {
        val state = state
        if (state != previousState) {
            if (Utils.isPlaying(state) && !Utils.isPlaying(previousState)) {
                manager.onPlay()
            } else if (Utils.isPaused(state) && !Utils.isPaused(previousState)) {
                manager.onPause()
            } else if (Utils.isStopped(state) && !Utils.isStopped(previousState)) {
                manager.onStop()
            }
            manager.onStateChange(state)
            previousState = state
            if (state == PlaybackStateCompat.STATE_STOPPED) {
                val previous = currentTrackIndex
                val position = position
                manager.onTrackUpdate(previous, position, null, null)
                manager.onEnd(currentTrackIndex, position)
            }
        }
    }

    override fun onPlayerError(error: ExoPlaybackException) {
        val code: String
        code = if (error.type == ExoPlaybackException.TYPE_SOURCE) {
            "playback-source"
        } else if (error.type == ExoPlaybackException.TYPE_RENDERER) {
            "playback-renderer"
        } else {
            "playback" // Other unexpected errors related to the playback
        }
        manager.onError(code, error.cause!!.message)
    }

    override fun onPlaybackParametersChanged(playbackParameters: PlaybackParameters) {
        // Speed or pitch changes
    }

    override fun onSeekProcessed() {
        // Finished seeking
    }

    override fun onMetadata(metadata: Metadata) {
        SourceMetadata.handleMetadata(manager, metadata)
    }

    init {
        val component = player!!.metadataComponent
        component?.addMetadataOutput(this)
    }
}